dive_core/gfxr_vulkan_command_hierarchy.cpp

```diff
@@ -0,0 +1,263 @@
+    m_capture_data_ptr = &capture_data;
+
+    // Clear/Reset internal data structures, just in case
+    *m_command_hierarchy_ptr = CommandHierarchy();
+
+    if (reserve_size.has_value())
+    {
+        for (uint32_t topology = 0; topology < (CommandHierarchy::kTopologyTypeCount); ++topology)
+        {
+            m_node_root_node_index[topology].reserve(*reserve_size);
+
+            m_node_children[topology][0].reserve(*reserve_size);
+            m_node_children[topology][1].reserve(*reserve_size);
+
+            m_command_hierarchy_ptr->m_nodes.m_node_type.reserve(*reserve_size);
+            m_command_hierarchy_ptr->m_nodes.m_description.reserve(*reserve_size);
+        }
+    }
+
+    // Add a dummy root node for easier management
+    uint64_t root_node_index = AddNode(NodeType::kRootNode, "", 0);
+    DIVE_VERIFY(root_node_index == Topology::kRootNodeIndex);
+
+    if (!ProcessGfxrSubmits(capture_data.GetGfxrSubmits(), capture_data.GetMemoryManager()))
+    {
+        return false;
+    }
+    // Convert the info in m_node_children into VulkanGfxrCommandHierarchy's topologies
+    CreateTopologies();
+
+    return true;
+}
+
+//--------------------------------------------------------------------------------------------------
+uint64_t GfxrVulkanCommandHierarchyCreator::AddNode(NodeType      type,
+                                                    std::string &&desc,
+                                                    char         *metadata_ptr,
+                                                    uint32_t      metadata_size)
+{
+    uint64_t node_index = m_command_hierarchy_ptr->AddGfxrNode(type, std::move(desc));
+
+    for (uint32_t i = 0; i < CommandHierarchy::kTopologyTypeCount; ++i)
+    {
+        DIVE_ASSERT(m_node_children[i][0].size() == node_index);
+        DIVE_ASSERT(m_node_children[i][1].size() == node_index);
+        m_node_children[i][0].resize(m_node_children[i][0].size() + 1);
+        m_node_children[i][1].resize(m_node_children[i][0].size() + 1);
+
+        m_node_root_node_index[i].resize(m_node_root_node_index[i].size() + 1);
+    }
+    return node_index;
+}
+
+//--------------------------------------------------------------------------------------------------
+void GfxrVulkanCommandHierarchyCreator::AddChild(CommandHierarchy::TopologyType type,
+                                                 uint64_t                       node_index,
+                                                 uint64_t                       child_node_index)
+{
+    DIVE_ASSERT(node_index < m_node_children[type][0].size());
+    m_node_children[type][0][node_index].push_back(child_node_index);
+}
+
+void GfxrVulkanCommandHierarchyCreator::GetArgs(const nlohmann::ordered_json &j,
+                                                uint64_t                      curr_index,
+                                                const std::string            &current_path)
+{
+    if (j.is_object())
```

Comment:
nit: i think it is clearer to use something like `json` as the var name, `j` seems a bit too similar as an iterator.

Reply:
Is json_args clearer?

Reply:
sure, much better, thanks :)